p4-16/spec/P4-16-spec.mdk

```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+integer value.
+
+~ Begin P4Example
+enum bit<8> NonUnique {
+  b1 = 0,
+  b2 = 1,  // Note, both b2 and b3 map to the same value.
+  b3 = 1,  
+  b4 = 2
+}
+~ End P4Example
+
+Additionally, within an `enum` declaration, earlier symbols may be referred to in later
+symbol declarations, so `NonUnique` could also be defined as follows.
+
+~ Begin P4Example
+enum bit<8> NonUnique {
+  b1 = 0,
+  b2 = 1,  // Note, both b2 and b3 map to the same value.
+  b3 = b2,  
+  b4 = 2
+}
+~ End P4Example
+
+An `enum` with an underlying type also supports explicit casts to and from the
+underlying type.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x;
+E a = E.e2;
+E b;
+
+x = (bit<8>) a; // sets x to 1
+b = (E) x;      // sets b to E.e2
+~ End P4Example
+
+casts `a`, which was initialized to `E.e2` to a `bit<8>`, using the specified
+fixed-width integer representation for `E.e2`, `1`.  The variable `b` is then set to the
+symbolic value `E.e2`, which corresponds to the fixed-width integer value `1`.
+
+Note that while it is always safe to cast from an `enum` to its fixed-width integer type,
+it may not be the case that every fixed-width integer value specified by the underlying type
+has a related symbol in the `enum`.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x = 5;
+E e = (E) x; // sets e to an unnamed value
+~ End P4Example
+
+sets `e` to an unnamed value, since there is no symbol corresponding to the
+fixed-width integer value `5`.  
+
+For example, in the following code, the `else` clause of the `if/else if/else`
+block can be reached even though the matches on `x` are complete with respect
+to the symbols defined in `MyPartialEnum_t`:
+
+~ Begin P4Example
+enum bit<2> MyPartialEnum_t {
+    VALUE_A = 2w0,
+    VALUE_B = 2w1,
+    VALUE_C = 2w2
+}
+
+bit<2> y = < some value >;
+MyPartialEnum_t x = (MyPartialEnum_t)y;
+
+if (x == MyPartialEnum_t.VALUE_A) {
+    // some code here
+} else if (x == MyPartialEnum_t.VALUE_B) {
+    // some code here
+} else if (x == MyPartialEnum_t.VALUE_C) {
+    // some code here
+} else {
+    // A P4 compiler MUST ASSUME that this branch can be executed
+    // some code here
+}
+~ End P4Example
+
+Additionally, if an enumeration is used as a field of a header, we would expect
+the `transition select` to match `default` when the parsed integer value does
+not match one of the symbolic values of `EtherType` in the following example:
+
+~ Begin P4Example
+enum bit<16> EtherType {
+  VLAN      = 0x8100,
+  IPV4      = 0x0800,
+  IPV6      = 0x86dd
+}
+
+header ethernet {
+  ...
+  EtherType etherType;
+  ...
+}
+
+parser my_parser(...) {
+  state parse_ethernet {
+    packet.extract(hdr.ethernet);
+    transition select(hdr.ethernet.etherType) {
+      EtherType.VLAN : parse_vlan;
+      EtherType.IPV4 : parse_ipv4;
+      EtherType.IPV6: parse_ipv6;
+      default: reject;
+  }
+}
+~ End P4Example
+
+Any `enum` that contains an unnamed value, whether as the result of a cast
```

Comment:
We should say here "variable with an enum type". The "enum" itself is the type declaration.

Reply:
Done.